network/common/src/main/java/org/apache/spark/network/buffer/WrappedLargeByteBuffer.java

```diff
@@ -0,0 +1,269 @@
+  private long _pos;
+  @VisibleForTesting
+  int currentBufferIdx;
+  @VisibleForTesting
+  ByteBuffer currentBuffer;
+
+
+  public WrappedLargeByteBuffer(ByteBuffer[] underlying) {
+    this(underlying, LargeByteBufferHelper.MAX_CHUNK);
+  }
+
+  /**
+   * you do **not** want to call this version.  It leads to a buffer which doesn't properly
+   * support {@link #asByteBuffer}.  The only reason it exists is to we can have tests which
+   * don't require 2GB of memory
+   *
+   * @param underlying
+   * @param subBufferSize
+   */
+  @VisibleForTesting
+  WrappedLargeByteBuffer(ByteBuffer[] underlying, int subBufferSize) {
+    if (underlying.length == 0) {
+      throw new IllegalArgumentException("must wrap at least one ByteBuffer");
+    }
+    this.underlying = underlying;
+    this.subBufferSize = subBufferSize;
+    long sum = 0L;
+    boolean startFound = false;
+    long initialPosition = -1;
+    for (int i = 0; i < underlying.length; i++) {
+      ByteBuffer b = underlying[i];
+      if (i != underlying.length -1 && b.capacity() != subBufferSize) {
+        throw new IllegalArgumentException("All buffers, except for the final one, must have " +
+          "size = " + subBufferSize);
+      }
+      if (startFound) {
+        if (b.position() != 0) {
+          throw new IllegalArgumentException("ByteBuffers have inconsistent positions");
+        }
+      } else if (b.position() != b.capacity()) {
+        startFound = true;
+        initialPosition = sum + b.position();
+      }
+      sum += b.capacity();
+    }
+    _pos = initialPosition;
+    currentBufferIdx = 0;
+    currentBuffer = underlying[0];
+    size = sum;
+  }
+
+  @Override
+  public void get(byte[] dest, int offset, int length) {
+    if (length > remaining()) {
+      throw new BufferUnderflowException();
+    }
+    int moved = 0;
+    while (moved < length) {
+      int toRead = Math.min(length - moved, currentBuffer.remaining());
+      currentBuffer.get(dest, offset + moved, toRead);
+      moved += toRead;
+      updateCurrentBuffer();
+    }
+    _pos += moved;
+  }
+
+  @Override
+  public LargeByteBuffer rewind() {
+    if (currentBuffer != null) {
+      currentBuffer.rewind();
+    }
+    while (currentBufferIdx > 0) {
+      currentBufferIdx -= 1;
+      currentBuffer = underlying[currentBufferIdx];
+      currentBuffer.rewind();
+    }
+    _pos = 0;
+    return this;
+  }
+
+  @Override
+  public WrappedLargeByteBuffer deepCopy() {
+    ByteBuffer[] dataCopy = new ByteBuffer[underlying.length];
+    for (int i = 0; i < underlying.length; i++) {
+      ByteBuffer b = underlying[i];
+      dataCopy[i] = ByteBuffer.allocate(b.capacity());
+      int originalPosition = b.position();
+      b.rewind();
+      dataCopy[i].put(b);
+      dataCopy[i].position(0);
+      b.position(originalPosition);
+    }
+    return new WrappedLargeByteBuffer(dataCopy, subBufferSize);
+  }
+
+  @Override
+  public byte get() {
+    byte r = currentBuffer.get();
```

Comment:
This will throw an NPE if it is called after you go past the last buffer. Docs say it should throw `BufferUnderflowException` instead.